assets/src/edit-story/components/popup/index.js

```diff
@@ -19,58 +19,168 @@
+  switch (placement) {
+    case 'bottom-start':
+    case 'top-start':
+    case 'left':
+    case 'left-start':
+    case 'left-end':
+      return bodyRect.left + anchorRect.left - spacing;
+    case 'bottom-end':
+    case 'top-end':
+    case 'right':
+    case 'right-start':
+    case 'right-end':
+      return bodyRect.left + anchorRect.left + anchorRect.width + spacing;
+    case 'bottom':
+    case 'top':
+      return bodyRect.left + anchorRect.left + anchorRect.width / 2;
+    default:
+      return 0;
+  }
+}
+
+function getYOffset(placement, spacing = 0, anchorRect) {
+  switch (placement) {
+    case 'bottom':
+    case 'bottom-start':
+    case 'bottom-end':
+    case 'left-end':
+    case 'right-end':
+      return anchorRect.top + anchorRect.height + spacing;
+    case 'top':
+    case 'top-start':
+    case 'top-end':
+    case 'left-start':
+    case 'right-start':
+      return anchorRect.top - spacing;
+    case 'right':
+    case 'left':
+      return anchorRect.top + anchorRect.height / 2;
+    default:
+      return 0;
+  }
+}
+
+function getOffset(placement, spacing, anchorRect, popupRect, bodyRect) {
+  const { height = 0 } = popupRect || {};
+  const { x: spacingH = 0, y: spacingV = 0 } = spacing || {};
+  // Horizontal
+  const offsetX = getXOffset(placement, spacingH, anchorRect, bodyRect);
+  const maxOffsetX = bodyRect.width - spacingH;
+  // Vertical
+  const offsetY = getYOffset(placement, spacingV, anchorRect);
+  const maxOffsetY = bodyRect.height - height - spacingV;
+  // Clamp values
+  return {
+    x: Math.max(0, Math.min(offsetX, maxOffsetX)),
+    y: Math.max(0, Math.min(offsetY, maxOffsetY)),
+  };
+}
+
+function Popup({ anchor, children, placement = 'bottom', spacing, isOpen }) {
   const [popupState, setPopupState] = useState(null);
-  const containerRef = useRef();
+  const containerRef = useRef(null);
 
-  useLayoutEffect(() => {
-    function positionPopup(evt) {
+  const positionPopup = useCallback(
+    (evt) => {
       // If scrolling within the popup, ignore.
-      if (evt && containerRef.current?.contains(evt.target)) {
+      if (evt && evt.target && containerRef.current?.contains(evt.target)) {
         return;
       }
+
       const anchorRect = anchor.current.getBoundingClientRect();
       const bodyRect = document.body.getBoundingClientRect();
+      const popupRect = containerRef.current?.getBoundingClientRect();
 
-      // Note: This displays the popup right under the node, currently no variations implemented.
-      // @todo Needs to display above the anchor instead if possible in case of not having room below!
       setPopupState({
-        width,
-        offset: {
-          x: bodyRect.right - anchorRect.right,
-          y: anchorRect.bottom,
-        },
+        offset: getOffset(placement, spacing, anchorRect, popupRect, bodyRect),
       });
-    }
-    positionPopup();
+    },
+    [containerRef, anchor, placement, spacing]
```

Comment:
It looks like you construct new `spacing` for each render. You might need to use `useMemo()` above to make it stable.